andela/Titan-Devs
# #163518705 Role based access control

#### What does this PR do?
Add role-based access control
#### Description of Task to be completed?
The task add the following functionality:
- Create, Read, Update and Delete(CRUD) on Role model
- Create, Read, Update and Delete(CRUD) on Permission model
- Assign Role to a User
- Remove a Role to a User
- Grant Permission to a Role
- Revoke a Permission to a Role
- Get Permission(s) given to a specific Role
- Get Role(s) given to a specific User 

#### How should this be manually tested?
*Clone the repo and check out `ft-role-based-access-control-163518705` branch
* Follow the instructions in ReadMe to setup the project
* Open Postman client
* EndPoints(API):
============
* Users
    *  `GET api/v1/users/:userId/roles` Get roles assigned to the specifi user
* Roles
   * `POST api/v1/roles` Create new role by providing `name` and `description`, name is required.
   * `POST api/v1/roles/:roleId/user/:userId` assign a role to a specific user
   * `GET api/v1/roles/:roleId` Get a specific role details.
   * `GET api/v1/roles` Get all roles.
   * `GET api/v1/roles/:roleId/permissions` Get all permissions assigned to a specific role.
   * `PUT api/v1/roles/:roleId` Update a Role by providing `name` and `description` name is required.
   * `DETELE api/v1/roles/:roleId` Delete a specific role.
   * `DETELE api/v1/roles/:id/remove` Remove a role to a specific user.
* Permissions
   * `POST api/v1/permissions` Create new permission by providing `name` and `description`, name is required.
   * `POST api/v1/permissions/:permissionId/role/:roleId` assign a permission to a specific Role
   * `GET api/v1/permissions/:permissionId` Get a specific permission details.
   * `GET api/v1/permissions` Get all permissions.
   * `PUT api/v1/permissions/:permissionId` Update a Permission by providing `name` and `description` name is required.
   * `DETELE api/v1/permissions/:permissionId` Delete a specific permission.
   * `DETELE api/v1/permission/:id/revoke` Revoke a permission to a specific role.
#### Any background context you want to provide?
A brief overview of how this task is implemented.
- Users can have multiple Roles assigned to them
- Roles can belong to many Users
- Permissions are assigned to the roles *instead* of users
- A Permission can belong to many roles
- A Role can have multiple Permissions
#### What are the relevant pivotal tracker stories?
- #163518705
